verifiers/internal/gcb/slsaprovenance/v1.0/provenance.go

```diff
@@ -0,0 +1,198 @@
+	}
+
+	if err := common.ValidateStatementTypes(provenance.StatementHeader.Type, provenance.StatementHeader.PredicateType, PredicateSLSAProvenance); err != nil {
+		return nil, fmt.Errorf("%w: %s", serrors.ErrorInvalidDssePayload, err.Error())
+	}
+	return &provenance, nil
+}
+
+func (p *Provenance) Predicate() (interface{}, error) {
+	return p.Pred, nil
+}
+
+func (p *Provenance) PredicateType() (string, error) {
+	return p.StatementHeader.PredicateType, nil
+}
+
+func (p *Provenance) Header() (intoto.StatementHeader, error) {
+	return p.StatementHeader, nil
+}
+
+// BuilderID implements Statement.BuilderID.
+func (p *Provenance) BuilderID() (string, error) {
+	return p.Pred.RunDetails.Builder.ID, nil
+}
+
+// BuildType implements Statement.BuildType.
+func (p *Provenance) BuildType() (string, error) {
+	return p.Pred.BuildDefinition.BuildType, nil
+}
+
+// GetSystemParameters implements Provenance.GetSystemParameters.
+func (p *Provenance) GetSystemParameters() (map[string]any, error) {
+	sysParams, ok := p.Pred.BuildDefinition.InternalParameters.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("%w: %s", serrors.ErrorInvalidDssePayload, "system parameters type")
+	}
+
+	return sysParams, nil
+}
+
+// SourceTag implements Provenance.SourceTag.
+func (p *Provenance) SourceTag() (string, error) {
+	sysParams, err := p.GetSystemParameters()
+	if err != nil {
+		return "", err
+	}
+	subsTagName, err := getSubstitutionsField(sysParams, "TAG_NAME")
+	if err != nil {
+		return "", err
+	}
+
+	extParams, err := p.externalParameters()
+	if err != nil {
+		return "", err
+	}
+
+	subsTag := "refs/tags/" + subsTagName
+	configTag, err := configSourceField(extParams, "ref")
+	if err != nil {
+		return "", err
+	}
+	if subsTag != configTag {
+		return "", fmt.Errorf("%w: '%v' != '%v'", serrors.ErrorInvalidDssePayload, subsTag, configTag)
+	}
+	return subsTagName, nil
+}
+
+func getSubstitutionsField(sysParams map[string]any, name string) (string, error) {
+	substitutions, ok := sysParams["systemSubstitutions"]
+	if !ok {
+		return "", fmt.Errorf("%w: no entry '%v' in substitution map", common.ErrSubstitution, "systemSubstitutions")
+	}
+	substitutionsMap, ok := substitutions.(map[string]interface{})
+	if !ok {
+		return "", fmt.Errorf("%w: no entry '%v' in substitution map", common.ErrSubstitution, "systemSubstitutions")
+	}
+	value, ok := substitutionsMap[name]
+	if !ok {
+		return "", fmt.Errorf("%w: no entry '%v' in substitution map", common.ErrSubstitution, name)
+	}
+	valueStr, ok := value.(string)
+	if !ok {
+		return "", fmt.Errorf("%w: value '%v' is not a string", common.ErrSubstitution, value)
+	}
+	return valueStr, nil
+}
+
+// SourceBranch implements Provenance.SourceBranch.
+func (p *Provenance) SourceBranch() (string, error) {
+	// NOTE: for the implementation, verify consistency between
+	// repository.ref and substitution's BRANCH. See SourceTag().
+	return "", fmt.Errorf("%w: branch verification", serrors.ErrorNotSupported)
```

Comment:
maybe?
```suggestion
	return "", fmt.Errorf("%w: branch verification not supported", serrors.ErrorNotSupported)
```

Reply:
ErrorNotSupported already shows "not supported".